SwiftCompilerSources/Sources/Optimizer/Utilities/WalkUtils.swift

```diff
@@ -0,0 +1,518 @@
+      }
+    case let ds as DestructureStructInst:
+      if let (index, path) = path.pop(kind: .structField) {
+        return walkDownUses(ofValue: ds.results[index], path: path, state: state)
+      } else if path.topMatchesAnyValueField {
+        return walkDownResults(of: ds, path: path, state: state)
+      } else {
+        return leafOrUnmatched(value: operand, path: path, state: state)
+      }
+    case let dt as DestructureTupleInst:
+      if let (index, path) = path.pop(kind: .tupleField) {
+        return walkDownUses(ofValue: dt.results[index], path: path, state: state)
+      } else if path.topMatchesAnyValueField {
+        return walkDownResults(of: dt, path: path, state: state)
+      } else {
+        return leafOrUnmatched(value: operand, path: path, state: state)
+      }
+    case is InitExistentialRefInst, is OpenExistentialRefInst,
+      is BeginBorrowInst, is CopyValueInst,
+      is UpcastInst, is UncheckedRefCastInst, is EndCOWMutationInst,
+      is RefToBridgeObjectInst, is BridgeObjectToRefInst:
+      return walkDownUses(ofValue: (instruction as! SingleValueInstruction), path: path, state: state)
+    case let mdi as MarkDependenceInst:
+      if operand.index == 0 {
+        return walkDownUses(ofValue: mdi, path: path, state: state)
+      } else {
+        return unmatchedPath(value: operand, path: path, state: state)
+      }
+    case let br as BranchInst:
+      let val = br.getArgument(for: operand)
+      if let (path, state) = shouldRecomputeDown(def: val, path: path, state: state) {
+        return walkDownUses(ofValue: val, path: path, state: state)
+      } else {
+        return false
+      }
+    case let cbr as CondBranchInst:
+      let val = cbr.getArgument(for: operand)
+      if let (path, state) = shouldRecomputeDown(def: val, path: path, state: state) {
+        return walkDownUses(ofValue: val, path: path, state: state)
+      } else {
+        return false
+      }
+    case let bcm as BeginCOWMutationInst:
+      return walkDownUses(ofValue: bcm.bufferResult, path: path, state: state)
+    default:
+      return leafUse(value: operand, path: path, state: state)
+    }
+  }
+  
+  /// Starts the walk
+  mutating func walkDownUses(ofValue: Value, path: Path, state: State) -> Bool {
+    for operand in ofValue.uses where !operand.isTypeDependent {
+      if walkDown(value: operand, path: path, state: state) {
+        return true
+      }
+    }
+    return false
+  }
+  
+  mutating func walkDownResults(of value: MultipleValueInstruction, path: Path, state: State) -> Bool {
+    for result in value.results {
+      if let (path, state) = shouldRecomputeDown(def: result, path: path, state: state) {
+        if walkDownUses(ofValue: result, path: path, state: state) {
+          return true
+        }
+      }
+    }
+    return false
+  }
+  
+  mutating func leafOrUnmatched(value: Operand, path: Path, state: State) -> Bool {
```

Comment:
The initial idea was that `leafUse` meant that either the use is a "direct use" (empty path) of the target value by an instruction handled by the walker or a transitive use, i.e. the using instruction is unhandled by the walker.

However I agree that this can be done by the client if necessary in `unmatchedPath`.